### PR TITLE
Add input validation to function 'CreateFundingStep'

### PR DESCRIPTION
**Pull request by jasonburkert-ccpo**
Adds input validation to function.  Also adds unit tests.  Input validation errors are included in a error_map included in error response per API spec.

Ticket: AT-6409

### CODE REVIEW

**Comment by jasonburkert-ccpo** [comment]
@kylelaker-ccpo @jeffsegal-ccpo Does this ValidationErrorResponse meet expections previously discussed?

**Comment by kylelaker-ccpo** [comment]
> @kylelaker-ccpo @jeffsegal-ccpo Does this `ValidationErrorResponse` meet expectations previously discussed?

LGTM!

**Comment by jasonburkert-ccpo** [comment]
I like that enthusiasm.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts [hunk 1]
```diff
@@ -18,41 +23,77 @@ export async function handler(event: APIGatewayProxyEvent): Promise<APIGatewayPr
...
+  // 1 d. Input validation must take place
+  //   v. CLIN value and obligated funds must be numbers
+  //   vi. CLIN value and obligated funds should be formatted as currency (i.e. “0.00”)
+
+  // 2 b. An error map must be returned
+
+  const clins = fundingStep.clins.values();
+  // All validation tests below are logically negating the expected state; i.e. if (!expected) { return validation error }
+  // Assertion messages are positive and describe the expected state.
+  for (const clin of clins) {
```
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts]: @kylelaker-ccpo The validation has been improved to collect all input validation errors from all clins within a Funding Step.  See ticket comments for details and examples.  Ready for re-review as of 090904693580707437fd42286a5f4251e5dcbb5a.

**Review by kylelaker-ccpo — CHANGES_REQUESTED** [review]
I haven't had coffee yet this AM so I think I had some thoughts out of order which doesn't help the flow of the GitHub comments. I also didn't run any of the code suggestions through an actual interpreter; they're all free-form in the GitHub UI. But in general:

- Let's try to cleanup `verror` to a type rather than mutable exported global state, whether it's a `type` or `interface`.
- Using `it.each` may reduce some repetition in the tests
- Every invocation of `Date.now()` returns a different value

I really appreciate how thorough the tests are!

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+  PATH_PARAMETER_REQUIRED_BUT_MISSING,
+  REQUEST_BODY_INVALID,
+  DATABASE_ERROR,
+  NO_SUCH_PORTFOLIO_DRAFT,
+} from "../../utils/errors";
+import { ProvisioningStatus } from "../../models/ProvisioningStatus";
+
+const yesterday = new Date(Date.now() - 24 * 60 * 60 * 1000).toISOString().slice(0, 10);
+const today = new Date(Date.now()).toISOString().slice(0, 10);
+const tomorrow = new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString().slice(0, 10);
```
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:30]: Good point about the call to Date.now() and the edge case.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts [hunk 3]
```diff
@@ -0,0 +1,418 @@
...
+  it("should return error when request body is empty", async () => {
+    const emptyRequest: APIGatewayProxyEvent = {
+      body: "",
+      pathParameters: { portfolioDraftId: uuidv4() },
+    } as any;
+    const result = await handler(emptyRequest);
+    expect(result).toBeInstanceOf(ErrorResponse);
+    expect(result).toEqual(REQUEST_BODY_EMPTY);
+    expect(result.statusCode).toEqual(ErrorStatusCode.BAD_REQUEST);
+    expect(JSON.parse(result.body).code).toEqual(ErrorCodes.OTHER);
```
Comment (jeffsegal-ccpo) [inline packages/api/portfolioDrafts/funding/createFundingStep.test.ts:99]: This is a redundant expectation right? Because if 

```
expect(result).toEqual(REQUEST_BODY_EMPTY);
```

passes, then we know what the HTTP Status code and error code will be right?

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+    const invalidBodyRequest: APIGatewayProxyEvent = {
+      body: JSON.stringify({ foo: "bar" }) + "}", // invalid json
+      pathParameters: { portfolioDraftId: uuidv4() },
+    } as any;
+    const result = await handler(invalidBodyRequest);
+    expect(result).toBeInstanceOf(ErrorResponse);
+    expect(result).toEqual(REQUEST_BODY_INVALID);
+    expect(result.statusCode).toEqual(ErrorStatusCode.BAD_REQUEST);
+    expect(JSON.parse(result.body).code).toEqual(ErrorCodes.OTHER);
+    expect(JSON.parse(result.body).message).toMatch(/A valid request body must be provided/);
```
Comment (jeffsegal-ccpo) [inline packages/api/portfolioDrafts/funding/createFundingStep.test.ts:112]: Ditto to the comment on redundant clauses here and likely a bunch of places.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts [hunk 5]
```diff
@@ -1,59 +1,98 @@
...
+export function validateFundingStepClins(fs: FundingStep): Array<typeof verror> {
+  const clins = fs.clins.values();
+  let errorAccumulator = Array<typeof verror>();
+  for (const clin of clins) {
+    const errors = validateClin(clin);
+    if (errors.length) {
+      errorAccumulator = [...errors, ...errorAccumulator];
+    }
+  }
+  return errorAccumulator;
```
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts:44]: I did try using `map` from your original suggestion many days ago, and couldn't make it work (don't recall the specific error).  Will see if it's possible here.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts [hunk 6]
```diff
@@ -1,59 +1,98 @@
...
+  isClin,
+  isValidUuidV4,
+} from "../../utils/validation";
+import { ErrorCodes } from "../../models/Error";
+
+/**
+ * Validation error tuple
+ * verror = ["clin_number", "param_name", "param_value", "message"];
+ */
+export let verror: [string, string, string, string];
```
[the remark this answers is not in the document]
Comment (jeffsegal-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts:28]: Along these lines, what are the best practices for sharing test code? It gets confusing easily - you don't want to put it in a `test` folder or name it `.test.ts` because that implies that it's literally testing something when in fact it's really "main" code that just doesn't ship with the product. The preferred solution in Java is to create a stand-alone jar which is then  depended upon at the `test` scope. But I don't see an obvious solution for us here that doesn't involve building our own modules, which feels like overkill for now.

And getting back to the task at hand - do we really want to `export` anything from a single unit test?

**Review by jeffsegal-ccpo — DISMISSED** [review]

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts (hunk 3 above)
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:99]: This is redundant and perhaps overly prescriptive, but I think it's good to include these expectations because the API spec explicitly states them for each combination of operation and response.  It's how I noticed we had `ErrorCodes.INVALID_INPUT` just about everywhere instead of `ErrorCodes.OTHER` and came to understand their intent.  I think there are still instances where INVALID_INPUT is returned incorrectly because the respective unit tests only check the response type.

Also, in some cases the API spec's stated response doesn't squarely fit one in `errors.ts` due to a less common combination of `statusCode` and `code` and `message` (aka response description).

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+}
+/**
+ * Returns a static clin containing bad inputs
+ * - obligated funds is greater than the total clin value
+ * @returns a complete Clin with values that should cause validation errors
+ */
+function mockClinObligatedGreaterThanTotal(): Clin {
+  return {
+    ...mockClin(),
+    ...{ obligated_funds: 2, total_clin_value: 1 },
```
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts]: Fixed this and all similar instances in 240b0e58b18056376f6a2c3f376e0b1b35c58d55.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts (hunk 3 above)
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:99]: For example, incorrect use of INVALID_INPUT can be seen [here](https://github.com/dod-ccpo/atat-web-api/blob/a261d6db9767bb8ed19959edb93e78c9dc525a85/packages/api/portfolioDrafts/portfolio/createPortfolioStep.ts#L13) and [here](https://github.com/dod-ccpo/atat-web-api/blob/a261d6db9767bb8ed19959edb93e78c9dc525a85/packages/api/portfolioDrafts/getPortfolioDrafts.ts#L10) and [here](https://github.com/dod-ccpo/atat-web-api/blob/a261d6db9767bb8ed19959edb93e78c9dc525a85/packages/api/portfolioDrafts/deletePortfolioDraft.ts#L19).

If the respective unit tests were checking the expected `statusCode`, and within the body, `code` and `message` per the API spec that would have eliminated that incorrect use.

Granted all of these examples are where an `ErrorResponse` is instantiated within the function rather than using one in `errors.ts`.  For any `ErrorResponse` from `errors.ts` that we've vetted and proven correct these expectation checks probably are redundant and unnecessary, but I still think they are useful while we're are building and shaking things out.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts [hunk 8]
```diff
@@ -63,3 +102,88 @@ export async function handler(event: APIGatewayProxyEvent): Promise<APIGatewayPr
...
+ * @returns a collection of four-element tuples containing clin number, property names and values that failed input validation, and a specific message
+ */
+export function validateClin(clin: unknown): Array<typeof verror> {
+  if (!isClin(clin)) {
+    throw Error("Input must be a Clin object");
+  }
+  const errors = Array<typeof verror>();
+  if (!isValidDate(clin.pop_start_date)) {
+    verror = [clin.clin_number, "pop_start_date", clin.pop_start_date, "start date must be a valid date"];
+    errors.push(verror);
```
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts]: Fixed in 9cc9529b61ec4ae05cb11e95aed5cd6970daa0a6.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts (hunk 3 above)
Comment (jeffsegal-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:99]: OK that's fair. And I believe it may not technically be redundant until Kyle's fix to tighten up the usages (so that all re-used error types do have the same values for the members that we'd expect like `code`). So I'm good to keep these as-is but eventually would think we ought to remove them so as not to confuse ourselves.

Comment (kylelaker-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:99]: #179 has been merged so it should not be possible to have an invalid code here unless you really go out of your way to do so in (which will hopefully be caught during a manual peer review)

**Comment by kylelaker-ccpo** [comment]
With #179 having been merged, it's likely that `packages/api/utils/response.ts`'s `ValidationErrorResponse`'s constructor is going to need to have the line `errorMap,` changed to `error_map: errorMap` when creating the `ValidationError` anonymous object. It's a little yuck but it lets us keep typical TypeScript formatting/style throughout the rest of the code base while limit snake_case to files that directly interact with `models/*`.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts]: `ClinValidationError` added to replace verror tuple in 13a0ea6903f2595a62d33296dfbca77f08c3ae83.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts (hunk 3 above)
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:99]: @kylelaker-ccpo Thanks for addressing INVALID_INPUT/OTHER issue more holistically in #179.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts [hunk 9]
```diff
@@ -63,3 +102,88 @@ export async function handler(event: APIGatewayProxyEvent): Promise<APIGatewayPr
...
+ * Validates the given clin object
+ * Returns a Clin-centric array of tuples representing validation errors.
+ * @param clin an object that looks like a Clin
+ * @returns a collection of four-element tuples containing clin number, property names and values that failed input validation, and a specific message
+ */
+export function validateClin(clin: unknown): Array<typeof verror> {
+  if (!isClin(clin)) {
+    throw Error("Input must be a Clin object");
+  }
+  const errors = Array<typeof verror>();
```
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts]: Addressed in 13a0ea6903f2595a62d33296dfbca77f08c3ae83.

File: packages/api/portfolioDrafts/funding/createFundingStep.ts [hunk 10]
```diff
@@ -63,3 +102,88 @@ export async function handler(event: APIGatewayProxyEvent): Promise<APIGatewayPr
...
+  });
+}
+
+/**
+ * Validates the given clin object
+ * Returns a Clin-centric array of tuples representing validation errors.
+ * @param clin an object that looks like a Clin
+ * @returns a collection of four-element tuples containing clin number, property names and values that failed input validation, and a specific message
+ */
+export function validateClin(clin: unknown): Array<typeof verror> {
```
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.ts]: Addressed in 13a0ea6903f2595a62d33296dfbca77f08c3ae83.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts (hunk 4 above)
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts:112]: Resolved in other comments.

**Comment by jasonburkert-ccpo** [comment]
@kylelaker-ccpo @jeffsegal-ccpo I believe all comments have been addressed.

File: packages/api/portfolioDrafts/funding/createFundingStep.test.ts [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+  it("should return error map entries when given Clin has nonsensical pop dates (start>end)", () => {
+    const errors = validateClin(mockClinStartAfterEnd());
+    expect(errors).toContainEqual(["0001", "pop_start_date", tomorrow, "start date must occur before end date"]);
+    expect(errors).toContainEqual(["0001", "pop_end_date", today, "start date must occur before end date"]);
+  });
+  it("should return error map entries when given Clin has nonsensical pop dates (start=end)", () => {
+    const errors = validateClin(mockClinStartEqualsEnd());
+    expect(errors).toContainEqual(["0001", "pop_start_date", today, "start date must occur before end date"]);
+    expect(errors).toContainEqual(["0001", "pop_end_date", today, "start date must occur before end date"]);
+  });
```
[the remark this answers is not in the document]
Comment (jasonburkert-ccpo) [reply packages/api/portfolioDrafts/funding/createFundingStep.test.ts]: Let's attempt this in a future exercise.

**Comment by jasonburkert-ccpo** [comment]
This is addressed in 88d67bc3b4b641d284f53f278a7d0b2b0c4ec175.

> With #179 having been merged, it's likely that `packages/api/utils/response.ts`'s `ValidationErrorResponse`'s constructor is going to need to have the line `errorMap,` changed to `error_map: errorMap` when creating the `ValidationError` anonymous object. It's a little yuck but it lets us keep typical TypeScript formatting/style throughout the rest of the code base while limit snake_case to files that directly interact with `models/*`.

File: packages/api/models/Error.ts [hunk 12]
```diff
@@ -9,6 +9,5 @@ export interface Error {
 }
 
 export interface ValidationError extends Error {
-  code: ErrorCode.INVALID_INPUT;
```
Comment (kylelaker-ccpo) [inline packages/api/models/Error.ts:12]: I think we want to keep `ErrorCode.INVALID_INPUT` correct? Because that enforces that ValidationErrors always have `INVALID_INPUT` for the `code`?

Comment (jasonburkert-ccpo) [reply packages/api/models/Error.ts:12]: I think I simply made a mistake attempting conflict resolution.

**Review by kylelaker-ccpo — APPROVED** [review]
